x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java

```diff
@@ -0,0 +1,100 @@
+import org.elasticsearch.index.reindex.UpdateByQueryAction;
+import org.elasticsearch.index.reindex.UpdateByQueryRequestBuilder;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptType;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+import static org.elasticsearch.xpack.core.ClientHelper.SECURITY_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_MAIN_ALIAS;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SecurityMainIndexMappingVersion.ADD_REMOTE_CLUSTER_AND_DESCRIPTION_FIELDS;
+
+public class SecurityMigrations {
+
+    public interface SecurityMigration {
+        /**
+         * Method that will execute the actual migration - needs to be idempotent and non-blocking
+         *
+         * @param securityIndex the main security index manager
+         * @param client the index client
+         * @param listener listener to provide updates back to caller
+         */
+        void migrate(SecurityIndexManager securityIndex, Client client, ActionListener<Void> listener);
+
+        /**
+         * Any node features that are required for this migration to run. This makes sure that all nodes in the cluster can handle any
+         * changes in behaviour introduced by the migration.
+         *
+         * @return a set of features needed to be supported or an empty set if no change in behaviour is expected
+         */
+        Set<NodeFeature> nodeFeaturesRequired();
+
+        /**
+         * The min mapping version required to support this migration. This makes sure that the index has at least the min mapping that is
+         * required to support the migration.
+         *
+         * @return the minimum mapping version required to apply this migration
+         */
+        int minMappingVersion();
+    }
+
+    public static final TreeMap<Integer, SecurityMigration> MIGRATIONS_BY_MIGRATION_VERSION = new TreeMap<>(
+        Map.of(1, new SecurityMigration() {
+            @Override
+            public void migrate(SecurityIndexManager securityIndex, Client client, ActionListener<Void> listener) {
+                UpdateByQueryRequestBuilder updateByQueryRequestBuilder = new UpdateByQueryRequestBuilder(client);
+                updateByQueryRequestBuilder.filter(new BoolQueryBuilder().should(QueryBuilders.termQuery("type", "role")));
+                updateByQueryRequestBuilder.source(SECURITY_MAIN_ALIAS);
+                updateByQueryRequestBuilder.script(
+                    new Script(
+                        ScriptType.INLINE,
+                        "painless",
+                        "ctx._source.metadata_flattened = ctx._source.metadata",
+                        Collections.emptyMap()
+                    )
+                );
+
+                securityIndex.checkIndexVersionThenExecute(
```

Comment:
I think the checks covered here should be included somewhere before the migration task is triggered (I haven't yet followed what happens if this check fails, or, more general, if the migration fails in which cases it is retried or not).

Reply:
Good point. I'll add that.